File: src/main/java/com/soapdogg/lispInterpreter/valueretriver/NumericValueRetriever.java
package com.soapdogg.lispInterpreter.valueretriver;

import com.soapdogg.lispInterpreter.determiner.NumericStringDeterminer;
import com.soapdogg.lispInterpreter.exceptions.NotAtomicException;
import com.soapdogg.lispInterpreter.exceptions.NotNumericException;
import com.soapdogg.lispInterpreter.printer.ListNotationPrinter;
import com.soapdogg.lispInterpreter.datamodels.Node;
import lombok.AllArgsConstructor;

@AllArgsConstructor(staticName = "newInstance")
public class NumericValueRetriever {

    private final AtomicValueRetriever atomicValueRetriever;
    private final NumericStringDeterminer numericStringDeterminer;
    private final ListNotationPrinter listNotationPrinter;

    public int retrieveNumericValue(
        final Node node,
        final int position,
        final String functionName
    ) throws NotNumericException, NotAtomicException {
        String value = atomicValueRetriever.retrieveAtomicValue(
            node,
            position,
            functionName
        );
        boolean isNumeric = numericStringDeterminer.isStringNumeric(value);
        if(!isNumeric) {
            String listNotation = listNotationPrinter.printInListNotation(
                node
            );
            String sb = "Error! Parameter at position: " + position +
                " of function " +
                functionName +
                " is not numeric!    Actual: " +
                listNotation +
                '\n';
            throw new NotNumericException(sb);
        }
        return Integer.parseInt(value);
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/valueretriver/NumericValueRetriever.kt
package com.soapdogg.lispInterpreter.valueretriver

import com.soapdogg.lispInterpreter.datamodels.AtomNode
import com.soapdogg.lispInterpreter.datamodels.NodeV2
import com.soapdogg.lispInterpreter.determiner.NumericStringDeterminer
import com.soapdogg.lispInterpreter.exceptions.NotNumericException
import com.soapdogg.lispInterpreter.printer.ListNotationPrinter

class NumericValueRetriever(
    private val numericStringDeterminer: NumericStringDeterminer,
    private val listNotationPrinter: ListNotationPrinter
) {

    fun retrieveNumericValue(
        node: NodeV2,
        functionName: String,
        index: Int
    ): Int {
        if (node is AtomNode) {
            val isNumeric = numericStringDeterminer.isStringNumeric(node.value)
            if (isNumeric) {
                return node.value.toInt()
            }
        }
        val value = listNotationPrinter.printInListNotation(node)
        val sb = """Error! Parameter at position: $index of function $functionName is not numeric!    Actual: $value${'\n'}"""
        throw NotNumericException(sb)
    }
}